packages/next/next-server/server/next-server.ts

```diff
@@ -1084,7 +1084,7 @@ export default class Server {
     let result: null | LoadComponentsReturnType = null
 
     // use static 404 page if available and is 404 response
-    if (this.nextConfig.experimental.static404 && err === null) {
+    if (this.nextConfig.experimental.static404 && res.statusCode === 404) {
```

Comment:
I think they both have to be checked:

```suggestion
    if (this.nextConfig.experimental.static404 && (res.statusCode === 404 || err === null)) {
```

Reply:
Rendering 404 when `err` is `null` seems to break the `Production Usage › should handle failed param decoding` test case

Reply:
@ijjk interesting! We should probably fix that later